Update cumulative to delta processor.

a-feld opened the pull request:
**Description:**

This PR improves on the cumulative to delta processor. Implemented in this PR:

* Resource and instrumentation library attributes are used as part of the metric identity (i.e. for two metrics with distinct resource attributes and the same metric name, the metrics will now aggregate separately)
* Detection of monotonic counter reset via start timestamp
* Detection of monotonic counter reset via non monotonic value
* Default to converting all monotonic cumulative counters
* Removes AWS metric library dependency

**Benchmarks:**

### Before this PR
```
/usr/local/bin/go test -benchmem -run=^$ -bench ^(BenchmarkConsumeMetrics)$ github.com/open-telemetry/opentelemetry-collector-contrib/processor/cumulativetodeltaprocessor

goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/processor/cumulativetodeltaprocessor
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkConsumeMetrics-12    	 1441018	       831.5 ns/op	     320 B/op	      11 allocs/op
```

### After this PR
```
/usr/local/bin/go test -benchmem -run=^$ -bench ^(BenchmarkConsumeMetrics)$ github.com/open-telemetry/opentelemetry-collector-contrib/processor/cumulativetodeltaprocessor

goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/processor/cumulativetodeltaprocessor
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkConsumeMetrics-12    	 2320220	       494.8 ns/op	     128 B/op	       5 allocs/op
PASS
ok  	github.com/open-telemetry/opentelemetry-collector-contrib/processor/cumulativetodeltaprocessor	1.903s
```

a-feld commented:
@hossain-rayhan @bogdandrutu This PR should be ready for review!
This is a follow up to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/4216#issuecomment-892222875

bogdandrutu reviewed (commented):
@a-feld this is a very large change 👍🏽 would be good to split it.

a-feld commented:
@bogdandrutu splitting things up is a great suggestion! Would it be OK to split this into multiple commits within this PR or would you prefer to have multiple PRs?

Additionally, if you have a preference for how to split this PR up, I would also be happy to follow any guidance 😄 
If not, I can work on splitting this out into logical smaller commits within the PR around the bullet points specified in the [summary](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/4444#issue-703901256).

a-feld committed:
Add metric tracking library to replace awsmetrics dependency.

a-feld committed:
Update configuration of cumulativetodelta processor.

The configuration has been updated so that:
* An empty list of metrics defaults to converting all cumulative metrics
  to delta.
* The state TTL (MaxStale) is now considered configurable
* There is a configuration option to enable/disable conversion of
  non-monotonic cumulative metrics. Non-monotonic cumulative metrics are
  normally considered gauges and aggregation temporality should not be
  converted. The default is that only monotonic values are converted.

a-feld committed:
Update cumulativetodelta processor logic.

* Uses the internal tracker library instead of AWS metrics library. This
  enables separation of timeseries by resource and instrumentation
  library.
* Metric data points which are invalid (the first in a series of
  non-monontic cumulative values) are now removed from the dataset.

a-feld committed:
Update processor default test case.

By default, the cumulative to delta processor now converts all metrics.
Previously, the processor did not convert any metrics unless listed in
the configuration.

a-feld committed:
Remove extra validate function call.

a-feld committed:
Add processor benchmark.

a-feld committed:
Remove aws metrics library.

a-feld commented:
Hi @bogdandrutu and @hossain-rayhan - I spent some time working to minimize any unnecessary changes to the processor. I have split the logic out into an ordered set of commits which should make review easier. Please let me know if there are any questions!

hossain-rayhan commented on processor/cumulativetodeltaprocessor/config.go:
```diff
@@ -24,15 +24,19 @@ import (
+	MaxStale time.Duration `mapstructure:"max_stale"`
+
+	// Set to false in order to convert non monotonic metrics
+	MonotonicOnly bool `mapstructure:"monotonic_only"`
```
What would be the default behavior here?

a-feld replied:
The default behavior is to convert only monotonic metrics. (defaults to true)
My understanding is that non-monotonic sum metrics are treated as gauges and therefore should not be converted by default 😄

hossain-rayhan commented on processor/cumulativetodeltaprocessor/config.go:
```diff
@@ -24,15 +24,19 @@ import (
 	Metrics []string `mapstructure:"metrics"`
+
+	// The total time a state entry will live past the time it was last seen. Set to 0 to retain state indefinitely.
+	MaxStale time.Duration `mapstructure:"max_stale"`
```
I am not sure about the use case here. Wondering, how important is it to expose as a config option? Less configuarability gives us simplicity.

a-feld replied:
The use case is for the prometheus receiver. The scrape interval is configurable and so we wouldn't want metrics to go stale prior to the scraper completing. Allowing this value to be configurable means that users can configure the staleness with their scrape interval + scrape duration + fudge factor.

jmacd replied:
Looks good. I suggest `MaxStaleness`.

hossain-rayhan commented on processor/cumulativetodeltaprocessor/config_test.go:
```diff
@@ -26,72 +26,45 @@ import (
-			configFile: "config_full.yaml",
 			expCfg: &Config{
-				ProcessorSettings: config.NewProcessorSettings(config.NewID(typeStr)),
+				ProcessorSettings: config.NewProcessorSettings(config.NewIDWithName(typeStr, "alt")),
```
nit: trying to learn. what advantage we are getting here using NewIDWithName() instead of NewID()?

hossain-rayhan replied:
Never mind. Got it.

hossain-rayhan commented on processor/cumulativetodeltaprocessor/processor.go:
```diff
@@ -16,32 +16,37 @@ package cumulativetodeltaprocessor
 
 type cumulativeToDeltaProcessor struct {
-	metrics         map[string]bool
+	metrics         map[string]struct{}
```
Can we rename this to `ConfiguredMetrics` or something like this? Also I suggestion I got from Bogdan is to write them with capital for exemplification.

hossain-rayhan commented on processor/cumulativetodeltaprocessor/factory.go:
```diff
@@ -42,11 +42,12 @@ func NewFactory() component.ProcessorFactory {
 func createDefaultConfig() config.Processor {
 	return &Config{
 		ProcessorSettings: config.NewProcessorSettings(config.NewID(typeStr)),
+		MonotonicOnly:     true,
```
Do we want to add a line in the README to the default value fo r`MonotoincOnly` is true?

a-feld committed:
Update readme.

a-feld committed:
Convert to using attributes rather than labels map.

a-feld commented:
Kindly pinging reviewers @hossain-rayhan @dashpole @Aneurysm9 @rakyll @bogdandrutu @alolita. We have folks waiting on this PR - your review would be appreciated! 😄

hossain-rayhan reviewed (approved):
LGTM. Thanks.

a-feld committed:
Fix usage of deprecated api.

a-feld committed:
Retain NaN values.

mtwo commented:
This needs to be merged before 9/24. @bogdandrutu will review.

dmitryax commented on processor/cumulativetodeltaprocessor/README.md:
```diff
@@ -9,18 +9,27 @@ The cumulative to delta processor (`cumulativetodeltaprocessor`) converts cumula
+
+- `metrics`: The processor uses metric names to identify a set of cumulative sum metrics and converts them to cumulative delta. Defaults to converting all metric names.
+- `max_stale`: The total time a state entry will live past the time it was last seen. Set to 0 to retain state indefinitely. Default: 0
+- `monotonic_only`: Specify whether only monotonic metrics are converted from cumulative to delta. Default: `true`. Set to `false` to convert metrics regardless of monotonic setting.
```
I there any real use case that you want to cover where the support of "gauges to deltas" needed? Otherwise I would recommend to remove this option and always skip non monotonic metrics.

a-feld replied:
True, this is a good point. I think I'll go ahead and remove this option. After that update, this library will only attempt to convert monotonic metrics.

dmitryax commented on processor/cumulativetodeltaprocessor/README.md:
```diff
@@ -9,18 +9,27 @@ The cumulative to delta processor (`cumulativetodeltaprocessor`) converts cumula
+
+The following settings can be optionally configured:
+
+- `metrics`: The processor uses metric names to identify a set of cumulative sum metrics and converts them to cumulative delta. Defaults to converting all metric names.
```
Similar question here. Do you have a use case where all metrics need to be translated? I think it would be the best if we will add regex support in future instead.

a-feld replied:
New Relic *requires* all metrics to be translated from cumulative -> delta temporality 😄 
The New Relic platform does not currently support cumulative metric temporalities.

jmacd commented on processor/cumulativetodeltaprocessor/config.go:
```diff
@@ -24,15 +24,19 @@ import (
+	// List of cumulative metrics to convert to delta. Default: converts all cumulative metrics to delta.
 	Metrics []string `mapstructure:"metrics"`
+
+	// The total time a state entry will live past the time it was last seen. Set to 0 to retain state indefinitely.
```
```suggestion
	// MaxStale is the total time a state entry will live past the time it was last seen. Set to 0 to retain state indefinitely.
```
style nit: Go style convention is that exported field comments should start with field name.

jmacd commented on processor/cumulativetodeltaprocessor/processor.go:
```diff
@@ -53,64 +58,92 @@ func (ctdp *cumulativeToDeltaProcessor) Start(context.Context, component.Host) e
-							fromDataPoint.SetStartTimestamp(pdata.NewTimestampFromTime(result.(delta).prevTimestamp))
-						}
-						metric.Sum().SetAggregationTemporality(pdata.AggregationTemporalityDelta)
+		ilms.RemoveIf(func(ilm pdata.InstrumentationLibraryMetrics) bool {
```
I prefer the new style here, thanks. 😀

jmacd commented on processor/cumulativetodeltaprocessor/processor.go:
```diff
@@ -53,64 +58,92 @@ func (ctdp *cumulativeToDeltaProcessor) Start(context.Context, component.Host) e
+			// the first data point is omitted since the initial
+			// reference is not assumed to be zero
+			if !valid {
+				return true
```
Sounds good. This behavior is described in https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#resets-and-gaps

jmacd commented on processor/cumulativetodeltaprocessor/tracking/identity.go:
```diff
@@ -0,0 +1,85 @@
+	MetricName             string
+	MetricUnit             string
+	StartTimestamp         pdata.Timestamp
+	Attributes             pdata.AttributeMap
```
This and the `Write` implementation make me nervous because Go has built-in support for using structs as map keys. The OTel-Go SDK exports an `attribute.Set` type used for metric identity inside the SDK, see https://github.com/open-telemetry/opentelemetry-go/blob/92551d3933c9c7ef5eaf4f93f876a5487d0024b9/attribute/set.go#L198

With one of the `attribute.Distinct` types for the resource and one for the attributes, the only other point type here is `InstrumentationLibrary`. I'd be happy with a TODO to revisit this.

jmacd commented on processor/cumulativetodeltaprocessor/tracking/tracker.go:
```diff
@@ -0,0 +1,191 @@
+	// NaN is used to signal "stale" metrics.
+	// These are ignored for now.
+	// https://github.com/open-telemetry/opentelemetry-collector/pull/3423
+	if metricID.IsFloatVal() && math.IsNaN(metricPoint.FloatValue) {
```
See also https://github.com/open-telemetry/opentelemetry-collector/issues/3987

jmacd commented on processor/cumulativetodeltaprocessor/tracking/tracker.go:
```diff
@@ -0,0 +1,191 @@
+		//	* Stale removal wins, updating goroutine will still see
+		//	  the removed state but the state after the update will
+		//	  not be persisted. The next update will load an entirely
+		//	  new state.
```
This sounds valid to me.

jmacd reviewed (approved):
This looks good. Given the circumstances of @a-feld's departure from the project, I advise merging this with a number of TODOs and associated tickets to continue ongoing development.

The only substantial point I've raised applies to the collector in general. There ought to be a standard way to create a map of metric identity, because more than one processor is going to use such a thing. The OTel-Go SDK has such a thing, but it's not perfectly simple to adapt. The tracking entry here could be replaced by a struct that is usable as a map key w/o the `Write()` mechanism, but I wouldn't hold up this PR because of it. The collector should support something like the OTel-Go `attribute.Distinct` type that applies to `model/pdata` types so that code like this can easily create a map key based on resource, instrumentation library, name, unit and attributes. I would assign the related issues to @hossain-rayhan and volunteer to help review.

jmacd commented:
This could be split into one part for staleness and one part for treating non-monotonic sums, but overall this looks like an improvement where benefits outweigh risks.

a-feld committed:
Exported field comments should start with field name.

Co-authored-by: Joshua MacDonald <jmacd@users.noreply.github.com>

a-feld committed:
Remove monotonic configuration option.

jmacd commented:
I hope someone at NR will pick this up!

alanwest commented:
> I hope someone at NR will pick this up!

@jmacd Yep! We'll work to bridge the gap soon. Thank you for your and others' extensive review to date.
